src/templates/TemplateData.ts

```diff
@@ -202,16 +231,56 @@ function getRuntimeKey(baseKey: string, runtime: ProjectRuntime): string {
     return runtime === ProjectRuntime.one ? baseKey : `${baseKey}.${runtime}`;
 }
 
-async function requestFunctionPortal<T>(hostname: string, subpath: string, runtime: string, param?: string): Promise<T> {
-    const options: request.OptionsWithUri = {
+async function cachedTemplatesAreCurrent(globalState?: vscode.Memento): Promise<boolean> {
+    if (globalState) {
+        for (const key of Object.keys(ProjectRuntime)) {
+            const runtime: ProjectRuntime = <ProjectRuntime>ProjectRuntime[key];
+            const funcJsonOptions: request.OptionsWithUri = {
+                method: 'GET',
+                uri: funcCliFeedUrl,
+                headers: {
+                    'User-Agent': 'Mozilla/5.0' // Required otherwise we get Unauthorized
+                }
+            };
+
+            const cliFeedJson: cliFeedJsonResponse = <cliFeedJsonResponse>JSON.parse(await <Thenable<string>>request(funcJsonOptions).promise());
+            const currentRelease: string = <string>cliFeedJson.tags[runtime].release;
+            const cachedRelease: string | undefined = globalState.get(`${runtime}-release`);
+            if (!cachedRelease || parseInt(cachedRelease, 10) < parseInt(currentRelease, 10)) {
+                // templates are not up-to-date and need to be downloaded/extracted
+                return false;
+            }
+
+            // make sure templates were cached
+            const cachedResources: object | undefined = globalState.get<object>(getRuntimeKey(resourcesKey, runtime));
+            const cachedTemplates: object[] | undefined = globalState.get<object[]>(getRuntimeKey(templatesKey, runtime));
+            const cachedConfig: object | undefined = globalState.get<object>(getRuntimeKey(configKey, runtime));
+            if (!cachedResources || !cachedTemplates || !cachedConfig) {
+                return false;
+            }
+        }
+        // only return true if ALL runtimes are current
+        return true;
+    }
+    return false;
+}
+
+async function getCurrentReleaseTemplates(runtime: string): Promise<string> {
```

Comment:
nit: I feel like getCurrentReleaseTemplates implies it would return the templates, not the version. How about "downloadLatestTemplates". I think that makes it more clear that the templates are a side-effect rather than the return value

Reply:
Should I make them two separate functions?

getCurrentReleaseVersion which calls downloadLatestTemplates?

Reply:
I think we should revisit this after you create a separate function to get cliFeedJson. You're getting it multiple times (once for each runtime in getCurrentReleaseTemplates and once in cachedTemplatesAreCurrent) and you only need to get it once and pass it along. That's actually most of the logic in this function, which means this function might not be necessary after your refactor. (In other words, you could move the remaining few lines directly into tryGetTemplateDataFromFuncPortal